### PR TITLE
prov/shm: Update SHM to use ROCR

### PR DESCRIPTION
**Pull request by amirshehataornl**
Update the SHM provider to use the ROCR HMEM asynchronous memory operations.

   . Add an IPC entry freestack
   . When progressing an IPC operation, check if the device is ROCR and trigger
     an asynchronous operation.
   . When an asynchronous operation is queued, create an IPC entry and
     add it to the queue of pending operations.
   . During the top level progress loop check the queue of pending asynchronous
     operations and query the state of each one. Generate a complete event
     for finished operations. Since completions happen outside the context
     of libfabric we can't rely on the ep->region->signal flag to be set.
     Always check the pending queue. This shouldn't introduce much of an
     overhead if the queue is empty

### CODE REVIEW

File: prov/shm/src/smr.h [hunk 1]
```diff
@@ -173,7 +175,8 @@ struct smr_cmd_ctx {
 OFI_DECLARE_FREESTACK(struct smr_rx_entry, smr_recv_fs);
 OFI_DECLARE_FREESTACK(struct smr_cmd_ctx, smr_cmd_ctx_fs);
 OFI_DECLARE_FREESTACK(struct smr_tx_entry, smr_pend_fs);
-OFI_DECLARE_FREESTACK(struct smr_sar_entry, smr_sar_fs);
+OFI_DECLARE_FREESTACK(struct smr_pend_entry, smr_sar_fs);
+OFI_DECLARE_FREESTACK(struct smr_pend_entry, smr_ipc_fs);
```
Comment (aingerson) [inline prov/shm/src/smr.h:179]: Since these are the same type and we should only ever need rx_size number of any of them, let's move this into one freestack (ie smr_pend_fs) and insert them into the same list (ie ep->rx_pend_list) that we can process differently based on the op. I think that will save memory and simplify the progressing code paths

File: prov/shm/src/smr_progress.c [hunk 2]
```diff
@@ -477,9 +477,69 @@ static struct smr_sar_entry *smr_progress_sar(struct smr_cmd *cmd,
...
+		   struct ofi_mr_entry *mr_entry, struct smr_cmd *cmd, int *rc)
+{
+	struct smr_pend_entry *ipc_entry;
+	enum fi_hmem_iface iface = cmd->msg.data.ipc_info.iface;
+	uint64_t device = cmd->msg.data.ipc_info.device;
+	int ret;
+
+	ipc_entry = ofi_freestack_pop(ep->ipc_pend_fs);
+	if (!ipc_entry)
+		assert(0);
```
Comment (aingerson) [inline prov/shm/src/smr_progress.c:492]: Let's handle this better with a warn and return

File: prov/shm/src/smr_progress.c [hunk 3]
```diff
@@ -519,6 +579,30 @@ static int smr_progress_ipc(struct smr_cmd *cmd, struct ofi_mr **mr,
...
+		if (!rx_entry) {
+			rx = calloc(sizeof(*rx), 1);
+			if (!rx) {
+				ret = -FI_ENOMEM;
+				goto out;
+			}
+
+			rx->iov = iov;
+			rx->count = iov_count;
+		}
```
Comment (aingerson) [inline prov/shm/src/smr_progress.c:596]: Why are we allocating an rx_entry every time? Just to save the iov/count? If we enter this if, that probably means we are doing an RMA which gets the iov from the cmd data instead of the rx_entry. The sar_entry struct has an iov/count for that reason. I don't think it is necessary to allocate a new one.

File: prov/shm/src/smr_progress.c [hunk 4]
```diff
@@ -519,6 +579,30 @@ static int smr_progress_ipc(struct smr_cmd *cmd, struct ofi_mr **mr,
 		ptr = (char *) (uintptr_t) mr_entry->info.ipc_mapped_addr +
 		      (uintptr_t) cmd->msg.data.ipc_info.offset;
 
+	if (cmd->msg.data.ipc_info.iface == FI_HMEM_ROCR) {
+		struct smr_pend_entry *ipc_entry;
```
Comment (aingerson) [inline prov/shm/src/smr_progress.c:583]: Pull out to the top even though it is only used for the if

File: prov/shm/src/smr_progress.c [hunk 5]
```diff
@@ -1062,18 +1149,65 @@ static void smr_progress_cmd(struct smr_ep *ep)
 	}
 }
 
+static void smr_progress_ipc_list(struct smr_ep *ep)
```
Comment (aingerson) [inline prov/shm/src/smr_progress.c:1152]: Merge with sar list progressing per first comment

File: prov/shm/src/smr_progress.c [hunk 6]
```diff
@@ -1062,18 +1149,65 @@ static void smr_progress_cmd(struct smr_ep *ep)
...
+{
+	struct smr_pend_entry *ipc_entry;
+	struct smr_domain *domain;
+	struct dlist_entry *tmp;
+	int ret;
+
+	domain = container_of(ep->util_ep.domain, struct smr_domain,
+			      util_domain);
+
+	pthread_spin_lock(&ep->region->lock);
```
Comment (aingerson) [inline prov/shm/src/smr_progress.c:1162]: Do we need to lock the whole region for this? Can we just use the util_ep lock like the sar processing does?

File: prov/shm/src/smr_progress.c [hunk 7]
```diff
@@ -1062,18 +1149,65 @@ static void smr_progress_cmd(struct smr_ep *ep)
...
+	/* after the synchronize all operations should be complete */
+	dlist_foreach_container_safe(&ep->ipc_cpy_pend_list,
+				     struct smr_pend_entry,
+				     ipc_entry, entry, tmp) {
+		enum fi_hmem_iface iface =
+			ipc_entry->cmd.msg.data.ipc_info.iface;
+		uint64_t device = ipc_entry->cmd.msg.data.ipc_info.device;
+
+		if (ofi_async_copy_query(iface, ipc_entry->async_event)
+				 != FI_SUCCESS)
```
Comment (aingerson) [inline prov/shm/src/smr_progress.c:1173]: Nit - move operand (!=) to first line before new line. Or I think you should just be able to check
`if (!ofi_async_copy_query(iface, ipc_entry->async_event))`

File: prov/util/src/rocr_mem_monitor.c [hunk 8]
```diff
@@ -37,6 +37,7 @@
 
 #include "ofi_tree.h"
 #include "ofi_iov.h"
+#include "ofi_hmem.h"
```
Comment (aingerson) [inline prov/util/src/rocr_mem_monitor.c:40]: This looks unrelated. Split into separate patch

**Review by aingerson — COMMENTED** [review]
Overall concept looks ok but would prefer some restructuring of the resources and progress

File: prov/shm/src/smr.h (hunk 1 above)
Comment (amirshehataornl) [reply prov/shm/src/smr.h:179]: The issue I see with this is that iterating through the ipc pend entries is fundamentally different from the sar entries. The completion of the async operations are external to the libfabric so it can't be gated by the signal. More over, I think it's cleaner to have separate lists for outstanding sar operations and ipc operations. They can use the same stack, but I believe the code is more straight forward if we queue them on separate lists. The processing functions of the operations is also different. I don't believe there is much of a memory overhead if we maintain the separate lists.

I'm okay with the rest of the comments. Will update.

File: prov/shm/src/smr_progress.c (hunk 5 above)
Comment (amirshehataornl) [reply prov/shm/src/smr_progress.c:1144]: as discussed in the earlier comment, it makes more sense to keep this separate.

**Comment by amirshehataornl** [comment]
@aingerson can I get another review on the PR please. And wondering what the failure in the continuous integration is? The DAOS test again?
Also when a PR has an "updates requested" tag, is there something I'm suppose to do? remove the tag? Doesn't look like I'm able to make any changes of that sort.

File: prov/shm/src/smr.h [hunk 9]
```diff
@@ -173,7 +175,7 @@ struct smr_cmd_ctx {
 OFI_DECLARE_FREESTACK(struct smr_rx_entry, smr_recv_fs);
 OFI_DECLARE_FREESTACK(struct smr_cmd_ctx, smr_cmd_ctx_fs);
 OFI_DECLARE_FREESTACK(struct smr_tx_entry, smr_pend_fs);
-OFI_DECLARE_FREESTACK(struct smr_sar_entry, smr_sar_fs);
```
Comment (aingerson) [inline prov/shm/src/smr.h:176]: Not sure what opr is? Can we call this rx_pend_fs or something to indicate that it is to track receive side pending operations? And change the tx one to tx_pend_fs?

File: prov/shm/src/smr.h [hunk 10]
```diff
@@ -145,6 +145,8 @@ struct smr_sar_entry {
 	size_t			iov_count;
 	struct ofi_mr		*mr[SMR_IOV_LIMIT];
 	bool			in_use;
+	struct ofi_mr_entry	*mr_entry;
```
Comment (aingerson) [inline prov/shm/src/smr.h:148]: Since this is only used for the IPC path to get the handle from the IPC cache, let's rename to make that clear so it's not confusing with the above ofi_mr data.

File: prov/shm/src/smr_progress.c [hunk 11]
```diff
@@ -519,6 +584,17 @@ static int smr_progress_ipc(struct smr_cmd *cmd, struct ofi_mr **mr,
 		ptr = (char *) (uintptr_t) mr_entry->info.ipc_mapped_addr +
 		      (uintptr_t) cmd->msg.data.ipc_info.offset;
 
+	if (cmd->msg.data.ipc_info.iface == FI_HMEM_ROCR) {
+		*total_len = 0;
+		ret = 0;
+		ipc_entry = smr_ipc_async_copy(ep,
+				(char*)ptr+cmd->msg.data.ipc_info.offset,
```
Comment (aingerson) [inline prov/shm/src/smr_progress.c:591]: Spaces around the +

File: prov/shm/src/smr_progress.c [hunk 12]
```diff
@@ -1062,18 +1141,79 @@ static void smr_progress_cmd(struct smr_ep *ep)
...
+
+	ofi_ep_lock_acquire(&ep->util_ep);
+
+	/* after the synchronize all operations should be complete */
+	dlist_foreach_container_safe(&ep->ipc_cpy_pend_list,
+				     struct smr_pend_entry,
+				     ipc_entry, entry, tmp) {
+		enum fi_hmem_iface iface =
+			ipc_entry->cmd.msg.data.ipc_info.iface;
+		uint64_t device = ipc_entry->cmd.msg.data.ipc_info.device;
```
Comment (aingerson) [inline prov/shm/src/smr_progress.c:1164]: Move the definitions to the beginning of the function

**Comment by aingerson** [comment]
@amirshehataornl CI failure is a compile error in the shm dsa code (see smr_dsa.c) - it accesses the sar entry directly so needs to be updated to use the pend_entry
prov/shm/src/smr_dsa.c:565:11: error: dereferencing pointer to incomplete type ‘struct smr_sar_entry’
sar_entry->bytes_done += dsa_cmd_context->bytes_in_progress;

File: prov/shm/src/smr.h (hunk 9 above)
Comment (amirshehataornl) [reply prov/shm/src/smr.h:176]: I renamed it to smr_tx_fs and smr_rx_fs respectively to remain consistent with the smr_ prepended to the other names. I also named the actual fields in the  smr_ep to tx_fs and rx_fs respectively

**Review by aingerson — APPROVED** [review]

File: prov/shm/src/smr.h [hunk 13]
```diff
@@ -172,8 +174,8 @@ struct smr_cmd_ctx {
 
 OFI_DECLARE_FREESTACK(struct smr_rx_entry, smr_recv_fs);
 OFI_DECLARE_FREESTACK(struct smr_cmd_ctx, smr_cmd_ctx_fs);
-OFI_DECLARE_FREESTACK(struct smr_tx_entry, smr_pend_fs);
-OFI_DECLARE_FREESTACK(struct smr_sar_entry, smr_sar_fs);
+OFI_DECLARE_FREESTACK(struct smr_tx_entry, smr_tx_fs);
+OFI_DECLARE_FREESTACK(struct smr_pend_entry, smr_rx_fs);
```
Comment (shefty) [inline prov/shm/src/smr.h:178]: It is confusing to see both a recv_fs and rx_fs.  It's doubly confusing to see that rx_entry is associated with the recv_fs, but tx_entry with tx_fs.

**Comment by shefty** [comment]
Once you're both happy with the changes, I'm fine.  Just the naming is very confusing to me.

File: prov/shm/src/smr.h (hunk 13 above)
Comment (aingerson) [reply prov/shm/src/smr.h:178]: that's a very good point. can we do a bit more refactoring and take something like what zach did to merge the tx and pend entry into one progress (or pend) entry?
https://github.com/ofiwg/libfabric/pull/8281/commits/e1720109269a23bc980dd5114dba4adeed83b914
Then we can keep rx_entry -> rx_fs
and then have one type of pend_fs but two different instances of them - rx_pend_fs and tx_pend_fs?

Comment (amirshehataornl) [reply prov/shm/src/smr.h:178]: Do you mean you want to do this refactoring in this patch? Wouldn't it make sense to add the refactoring in a separate PR as this is not related to the target of this PR?

Comment (aingerson) [reply prov/shm/src/smr.h:178]: I was suggesting adding a patch within this series, between the first and the async one, to merge them but I'm ok with doing that as a follow up if you prefer

Comment (shefty) [reply prov/shm/src/smr.h:178]: For this patch, maybe just rename rx_fs to pend_fs?

Comment (amirshehataornl) [reply prov/shm/src/smr.h:178]: If you guys are okay with it, I'd like to rename the rx_fs->pend_fs as Sean suggested for this PR, then I can touch base with Alexia to more fully understand the refactoring she's suggesting. I don't mind creating a separate PR for that.

Comment (aingerson) [reply prov/shm/src/smr.h:178]: That's fine with me. Thanks!

**Comment by amirshehataornl** [comment]
bot:aws:retest

**Comment by amirshehataornl** [comment]
@aingerson, Does the PR look good now?
